packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -3,30 +3,365 @@
+        uint256 _vaultId,
+        uint256 _wPowerPerpAmount,
+        uint256 _collateralAmount
+    ) external payable {
+        uint256 amountToFlashswap = _collateralAmount.sub(msg.value);
+
+        _exactInFlashSwap(
+            wPowerPerp,
+            weth,
+            IUniswapV3Pool(wPowerPerpPool).fee(),
+            _wPowerPerpAmount,
+            amountToFlashswap,
+            uint8(FLASH_SOURCE.FLASH_W_MINT),
+            abi.encodePacked(_vaultId, amountToFlashswap, _collateralAmount, _wPowerPerpAmount)
+        );
+
+        emit FlashswapWMint(msg.sender, _vaultId, _wPowerPerpAmount, amountToFlashswap, _collateralAmount);
+    }
+
+    /**
+     * @notice flash close position and buy long squeeth
+     * @dev this function
+     * @param _vaultId vault ID
+     * @param _wPowerPerpAmountToBurn amount of WPowerPerp to burn
+     * @param _wPowerPerpAmountToBuy amount of WPowerPerp to buy
+     * @param _collateralToWithdraw amount of collateral to withdraw from vault
+     * @param _collateralToBuyWith amount of collateral from vault to use to buy long
+     */
+    function flashswapWBurnBuyLong(
+        uint256 _vaultId,
+        uint256 _wPowerPerpAmountToBurn,
+        uint256 _wPowerPerpAmountToBuy,
+        uint256 _collateralToWithdraw,
+        uint256 _collateralToBuyWith
+    ) external payable {
+        require(_collateralToBuyWith <= _collateralToWithdraw.add(msg.value), "Not enough collateral");
+
+        _exactOutFlashSwap(
+            weth,
+            wPowerPerp,
+            IUniswapV3Pool(wPowerPerpPool).fee(),
+            _wPowerPerpAmountToBurn.add(_wPowerPerpAmountToBuy),
+            _collateralToBuyWith.add(msg.value),
+            uint8(FLASH_SOURCE.FLASH_W_BURN),
+            abi.encodePacked(
+                _vaultId,
+                _wPowerPerpAmountToBurn,
+                _wPowerPerpAmountToBuy,
+                _collateralToWithdraw,
+                _collateralToBuyWith.add(msg.value)
+            )
+        );
+
+        emit FlashWBurn(msg.sender, _vaultId, _wPowerPerpAmountToBurn, _collateralToWithdraw, _wPowerPerpAmountToBuy);
+    }
+
+    /**
+     * @notice sell long wPowerPerp and flashswap mint short position
+     * @dev flahswap amount = collateral amount - msg.value - ETH from selling long wPowerPerp
+     * @param _vaultId vault ID
+     * @param _wPowerPerpAmountToMint wPowerPerp amount to mint
+     * @param _collateralAmount collateral amount to use for minting
+     * @param _wPowerPerpAmountToSell long wPowerPerp amount to sell
+     * @param _minToReceive min ETH amount to receive for selling long wPowerPerp amount
+     */
+    function flashswapSellLongWMint(
+        uint256 _vaultId,
+        uint256 _wPowerPerpAmountToMint,
+        uint256 _collateralAmount,
+        uint256 _wPowerPerpAmountToSell,
+        uint256 _minToReceive
+    ) external payable {
+        IWPowerPerp(wPowerPerp).transferFrom(msg.sender, address(this), _wPowerPerpAmountToSell);
```

Comment:
should we have a require here (may just fail earlier if incorrectly sent tx vs waiting to fail in the callback so probably not strictly necessary)?

```require(_minToReceive.add(msg.value)>=_collateralAmount)```